File: app/src/main/java/com/artemchep/horario/database/models/Subject.java
/*
 * Copyright (C) 2017 XJSHQ@github.com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA  02110-1301, USA.
 */
package com.artemchep.horario.database.models;

import android.os.Parcel;
import android.support.annotation.NonNull;

import com.artemchep.horario.models.Model;

import org.apache.commons.lang.builder.EqualsBuilder;
import org.apache.commons.lang.builder.HashCodeBuilder;

/**
 * @author Artem Chepurnoy
 */
public class Subject extends Model {

    public static final Creator<Subject> CREATOR = new Creator<Subject>() {
        @Override
        public Subject[] newArray(int size) {
            return new Subject[size];
        }

        @Override
        public Subject createFromParcel(Parcel source) {
            return new Subject(source);
        }
    };

    public String data;  // key
    public String group;  // key

    public String name;
    public String abbreviation;
    /**
     * Should be one of those
     * {@link com.artemchep.horario.Palette#PALETTE colors}.
     */
    public int color;

    public Subject() {
        // Default constructor required for calls
        // to DataSnapshot.getValue(Subject.class)
    }

    public Subject(Parcel source) {
        key = source.readString();
        data = source.readString();
        group = source.readString();
        name = source.readString();
        abbreviation = source.readString();
        color = source.readInt();
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public int hashCode() {
        return new HashCodeBuilder(247, 11)
                .append(key)
                .append(group)
                .append(data)
                .append(name)
                .append(abbreviation)
                .append(color)
                .toHashCode();
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public boolean equals(Object o) {
        if (o == null) return false;
        if (o == this) return true;
        if (!(o instanceof Subject)) return false;

        Subject subject = (Subject) o;
        return new EqualsBuilder()
                .append(color, subject.color)
                .append(key, subject.key)
                .append(data, subject.data)
                .append(group, subject.group)
                .append(name, subject.name)
                .append(abbreviation, subject.abbreviation)
                .isEquals();
    }

    @NonNull
    public Subject clone() {
        return (Subject) super.clone();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(key);
        dest.writeString(data);
        dest.writeString(group);
        dest.writeString(name);
        dest.writeString(abbreviation);
        dest.writeInt(color);
    }

}

File: app/src/main/java/com/artemchep/horario/database/data/Subject.kt
package com.artemchep.horario.database.data

import android.os.Parcel
import com.artemchep.horario.Palette
import com.artemchep.horario.database.createParcel
import com.artemchep.horario.database.models.ISubject
import com.artemchep.horario.extensions.toBoolean
import com.artemchep.horario.extensions.toInt
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.Exclude
import com.google.firebase.firestore.DocumentSnapshot

/**
 * @author Artem Chepurnoy
 */
data class Subject(
        @Exclude
        override var id: String? = null,
        override var name: String? = null,
        override var author: String? = null,
        override var description: String? = null,
        override var archived: Boolean = false,
        override var color: Int = 0) : ISubject {

    companion object {

        @JvmField
        val CREATOR = createParcel { Subject(it) }

        fun clone(subject: Subject?) = subject?.copy()

        fun from(snapshot: DocumentSnapshot): Subject? {
            return if (snapshot.exists()) {
                Subject(id = snapshot.id).apply {
                    name = snapshot.getString("name")
                    author = snapshot.getString("author")
                    description = snapshot.getString("description")
                    color = snapshot.getLong("color")?.toInt() ?: Palette.GREY
                }
            } else null
        }

        fun from(snapshot: DataSnapshot): Subject? {
            return Subject(id = snapshot.key).apply {
                name = snapshot.child("name").value as String?
                author = snapshot.child("author").value as String?
                description = snapshot.child("description").value as String?
                archived = snapshot.child("archived").value.let { it as? Boolean } == true
                color = snapshot.child("color").value.let { it as? Long }?.toInt() ?: 0
            }
        }

        fun delta(a: Subject, b: Subject): HashMap<String, Any?> {
            return HashMap<String, Any?>().apply {
                if (a.name != b.name) put("name", a.name)
                if (a.author != b.author) put("author", a.author)
                if (a.description != b.description) put("description", a.description)
                if (a.archived != b.archived) put("archived", a.archived)
                if (a.color != b.color) put("color", a.color)
            }
        }

    }

    constructor(id: String?) : this(id, null)

    constructor(src: Parcel) : this(
            id = src.readString(),
            name = src.readString(),
            author = src.readString(),
            description = src.readString(),
            archived = src.readInt().toBoolean(),
            color = src.readInt()
    )

    override fun getKey(): String = id!!

    override fun describeContents(): Int = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.apply {
            writeString(id)
            writeString(name)
            writeString(author)
            writeString(description)
            writeInt(archived.toInt())
            writeInt(color)
        }
    }

}